File: maven/src/main/java/com/antwerkz/critter/UpdaterBuilder.java
package com.antwerkz.critter;

import com.mongodb.WriteConcern;
import com.mongodb.WriteResult;
import static java.lang.String.format;
import org.jboss.forge.roaster.Roaster;
import org.jboss.forge.roaster.model.source.FieldSource;
import org.jboss.forge.roaster.model.source.JavaClassSource;
import org.jboss.forge.roaster.model.source.MethodSource;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.query.UpdateOperations;
import org.mongodb.morphia.query.UpdateResults;

public class UpdaterBuilder {
  public UpdaterBuilder(final CritterClass critterClass, final JavaClassSource criteriaClass) {
    final String type = critterClass.getName() + "Updater";
    final MethodSource<JavaClassSource> getUpdater = criteriaClass.addMethod()
        .setPublic()
        .setName("getUpdater")
        .setReturnType(type);
    getUpdater
        .setBody(format("return new %s();", getUpdater.getReturnType()));

    final JavaClassSource updater = Roaster.create(JavaClassSource.class)
        .setPublic()
        .setName(type);

    criteriaClass.addImport(UpdateOperations.class);
    criteriaClass.addImport(UpdateResults.class);
    criteriaClass.addImport(WriteConcern.class);
    criteriaClass.addImport(WriteResult.class);

    final FieldSource<JavaClassSource> updateOperations = updater.addField()
        .setType(format("UpdateOperations<%s>", critterClass.getName()))
        .setLiteralInitializer(format("ds.createUpdateOperations(%s.class);", critterClass.getName()));
    updateOperations.setName("updateOperations");

    updater.addMethod()
        .setPublic()
        .setName("updateAll")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.update(query(), updateOperations, false);");

    updater.addMethod()
        .setPublic()
        .setName("updateFirst")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.updateFirst(query(), updateOperations, false);");

    updater.addMethod()
        .setPublic()
        .setName("updateAll")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.update(query(), updateOperations, false, wc);")
        .addParameter(WriteConcern.class, "wc");

    updater.addMethod()
        .setPublic()
        .setName("updateFirst")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.updateFirst(query(), updateOperations, false, wc);")
        .addParameter(WriteConcern.class, "wc");

    updater.addMethod()
        .setPublic()
        .setName("upsert")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.update(query(), updateOperations, true);");

    updater.addMethod()
        .setPublic()
        .setName("upsert")
        .setReturnType(UpdateResults.class)
        .setBody("return ds.update(query(), updateOperations, true, wc);")
        .addParameter(WriteConcern.class, "wc");

    updater.addMethod()
        .setPublic()
        .setName("remove")
        .setReturnType(WriteResult.class)
        .setBody("return ds.delete(query());");

    updater.addMethod()
        .setPublic()
        .setName("remove")
        .setReturnType(WriteResult.class)
        .setBody("return ds.delete(query(), wc);")
        .addParameter(WriteConcern.class, "wc");

    for (CritterField field : critterClass.getFields()) {
      if (!field.getParameterTypes().isEmpty()) {
        field.getParameterTypes()
            .stream()
            .forEach(criteriaClass::addImport);
      }

      criteriaClass.addImport(field.getFullType());
      if (!field.hasAnnotation(Id.class)) {
        updater.addMethod()
            .setPublic()
            .setName(field.getName())
            .setReturnType(type)
            .setBody(format("updateOperations.set(\"%s\", value);\nreturn this;", field.getName()))
            .addParameter(field.getParameterizedType(), "value");

        updater.addMethod()
            .setPublic()
            .setName(format("unset%s", nameCase(field.getName())))
            .setReturnType(type)
            .setBody(format("updateOperations.unset(\"%s\");\nreturn this;", field.getName()));

        numerics(type, updater, field);
        containers(type, updater, field);
      }
    }

    criteriaClass.addNestedType(updater);
  }

  private void numerics(final String type, final JavaClassSource updater, final CritterField field) {
    if(field.isNumeric()) {
      updater.addMethod()
          .setPublic()
          .setName(format("dec%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.dec(\"%s\");\nreturn this;", field.getName()));

      updater.addMethod()
          .setPublic()
          .setName(format("inc%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.inc(\"%s\");\nreturn this;", field.getName()));


      updater.addMethod()
          .setPublic()
          .setName(format("inc%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.inc(\"%s\", value);\nreturn this;", field.getName()))
          .addParameter(field.getFullType(), "value");
    }
  }
  private void containers(final String type, final JavaClassSource updater, final CritterField field) {
    if(field.isContainer()) {

      updater.addMethod()
          .setPublic()
          .setName(format("addTo%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.add(\"%s\", value);\nreturn this;", field.getName()))
          .addParameter(field.getParameterizedType(), "value");

      MethodSource<JavaClassSource> addItems = updater.addMethod()
          .setPublic()
          .setName(format("addTo%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.add(\"%s\", value, addDups);\nreturn this;", field.getName()));
      addItems
          .addParameter(field.getParameterizedType(), "value");
      addItems
          .addParameter("boolean", "addDups");

      addItems = updater.addMethod()
          .setPublic()
          .setName(format("addAllTo%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.addAll(\"%s\", values, addDups);\nreturn this;", field.getName()));
      addItems.addParameter(field.getParameterizedType(), "values");
      addItems.addParameter("boolean", "addDups");

      updater.addMethod()
          .setPublic()
          .setName(format("removeFirstFrom%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.removeFirst(\"%s\");\nreturn this;", field.getName()));

      updater.addMethod()
          .setPublic()
          .setName(format("removeLastFrom%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.removeLast(\"%s\");\nreturn this;", field.getName()));

      updater.addMethod()
          .setPublic()
          .setName(format("removeFrom%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.removeAll(\"%s\", value);\nreturn this;", field.getName()))
          .addParameter(field.getParameterizedType(), "value");

      MethodSource<JavaClassSource> removeAll = updater.addMethod()
          .setPublic()
          .setName(format("removeAllFrom%s", nameCase(field.getName())))
          .setReturnType(type)
          .setBody(format("updateOperations.removeAll(\"%s\", values);\nreturn this;", field.getName()));
      removeAll.addParameter(field.getParameterizedType(), "values");
    }
  }

  private String nameCase(final String name) {
    return name.substring(0, 1).toUpperCase() + name.substring(1);
  }
}

File: maven/src/main/kotlin/com/antwerkz/critter/UpdaterBuilder.kt
package com.antwerkz.critter

import com.mongodb.WriteConcern
import com.mongodb.WriteResult
import org.mongodb.morphia.annotations.Id
import org.mongodb.morphia.query.UpdateOperations
import org.mongodb.morphia.query.UpdateResults

class UpdaterBuilder(sourceClass: CritterClass, targetClass: CritterClass) {
    init {
        val type = sourceClass.getName() + "Updater"
        val method = targetClass.addMethod()
                .setPublic()
                .setName("getUpdater")
                .setReturnType(type)
        method.setBody("return new ${method.getReturnType()}();")

        val updater = targetClass.createClass(name = type)

        targetClass.addImport(UpdateOperations::class.java)
        targetClass.addImport(UpdateResults::class.java)
        targetClass.addImport(WriteConcern::class.java)
        targetClass.addImport(WriteResult::class.java)

        updater.addField("updateOperations", "UpdateOperations<${sourceClass.getName()}>")
                .setLiteralInitializer("ds.createUpdateOperations(${sourceClass.getName()}.class);")

        updater.addMethod()
                .setPublic()
                .setName("updateAll")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.update(query(), updateOperations, false);")

        updater.addMethod()
                .setPublic()
                .setName("updateFirst")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.updateFirst(query(), updateOperations, false);")

        updater.addMethod()
                .setPublic()
                .setName("updateAll")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.update(query(), updateOperations, false, wc);")
                .addParameter(WriteConcern::class.java, "wc")

        updater.addMethod()
                .setPublic()
                .setName("updateFirst")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.updateFirst(query(), updateOperations, false, wc);")
                .addParameter(WriteConcern::class.java, "wc")

        updater.addMethod()
                .setPublic()
                .setName("upsert")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.update(query(), updateOperations, true);")

        updater.addMethod()
                .setPublic()
                .setName("upsert")
                .setReturnType(UpdateResults::class.java)
                .setBody("return ds.update(query(), updateOperations, true, wc);")
                .addParameter(WriteConcern::class.java, "wc")

        updater.addMethod()
                .setPublic()
                .setName("remove")
                .setReturnType(WriteResult::class.java)
                .setBody("return ds.delete(query());")

        updater.addMethod()
                .setPublic()
                .setName("remove")
                .setReturnType(WriteResult::class.java)
                .setBody("return ds.delete(query(), wc);")
                .addParameter(WriteConcern::class.java, "wc")

        sourceClass.fields
                .filter({ field -> !field.isStatic() })
                .forEach { field ->
                    if (!field.parameterTypes.isEmpty()) {
                        field.parameterTypes
                                .forEach { targetClass.addImport(it) }
                    }

                    targetClass.addImport(field.fullType)
                    if (!field.hasAnnotation(Id::class.java)) {
                        updater.addMethod()
                                .setPublic()
                                .setName(field.name)
                                .setReturnType(type)
                                .setBody("updateOperations.set(\"${field.name}\", value);\nreturn this;")
                                .addParameter(field.parameterizedType, "value")

                        updater.addMethod()
                                .setPublic()
                                .setName("unset${nameCase(field.name)}")
                                .setReturnType(type)
                                .setBody("updateOperations.unset(\"${field.name}\");\nreturn this;")

                        numerics(type, updater, field)
                        containers(type, updater, field)
                    }
                }
        targetClass.addNestedType(updater)
    }

    private fun numerics(type: String, updater: CritterClass, field: CritterField) {
        if (field.isNumeric()) {
            updater.addMethod()
                    .setPublic()
                    .setName("dec${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.dec(\"${field.name}\");\nreturn this;")

            updater.addMethod()
                    .setPublic()
                    .setName("inc${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.inc(\"${field.name}\");\nreturn this;")

            updater.addMethod()
                    .setPublic()
                    .setName("inc${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.inc(\"${field.name}\", value);\nreturn this;")
                    .addParameter(field.fullType, "value")
        }
    }

    private fun containers(type: String, updater: CritterClass, field: CritterField) {
        if (field.isContainer()) {

            updater.addMethod()
                    .setPublic()
                    .setName("addTo${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.add(\"${field.name}\", value);\nreturn this;")
                    .addParameter(field.parameterizedType, "value")

            updater.addMethod()
                    .setPublic()
                    .setName("addTo${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.add(\"${field.name}\", value, addDups);\nreturn this;")
                    .addParameter(field.parameterizedType, "value")
                    .addParameter("boolean", "addDups")

            updater.addMethod()
                    .setPublic()
                    .setName("addAllTo${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.addAll(\"${field.name}\", values, addDups);\nreturn this;")
                    .addParameter(field.parameterizedType, "values")
                    .addParameter("boolean", "addDups")

            updater.addMethod()
                    .setPublic()
                    .setName("removeFirstFrom${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.removeFirst(\"${field.name}\");\nreturn this;")

            updater.addMethod()
                    .setPublic()
                    .setName("removeLastFrom${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.removeLast(\"${field.name}\");\nreturn this;")

            updater.addMethod()
                    .setPublic()
                    .setName("removeFrom${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.removeAll(\"${field.name}\", value);\nreturn this;")
                    .addParameter(field.parameterizedType, "value")

            updater.addMethod()
                    .setPublic()
                    .setName("removeAllFrom${nameCase(field.name)}")
                    .setReturnType(type)
                    .setBody("updateOperations.removeAll(\"${field.name}\", values);\nreturn this;")
                    .addParameter(field.parameterizedType, "values")
        }
    }

    private fun nameCase(name: String): String {
        return name.substring(0, 1).toUpperCase() + name.substring(1)
    }
}
